server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -57,12 +60,56 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
 
     final Map<String, Integer> loadStatus = Maps.newHashMap();
 
+    final Map<String, Integer> tieredReplicants = getTieredReplicants();
+    for (final String tier : tieredReplicants.keySet()) {
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, 0);
+    }
+
+    final BalancerStrategy strategy = params.getBalancerStrategy();
+
+    final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
+                                                    .getMaxSegmentsInNodeLoadingQueue();
+
+    final Predicate<ServerHolder> serverHolderPredicate;
+    if (maxSegmentsInNodeLoadingQueue > 0) {
+      serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
+    } else {
+      serverHolderPredicate = Objects::nonNull;
+    }
+
     int totalReplicantsInCluster = params.getSegmentReplicantLookup().getTotalReplicants(segment.getIdentifier());
-    for (Map.Entry<String, Integer> entry : getTieredReplicants().entrySet()) {
+
+    final Optional<String> primaryLoadingTier;
+    if (totalReplicantsInCluster <= 0) {
```

Comment:
I would also add trace logging to help debugging the issue in the future. Usually a log for each important if branch is good, which helps to understand which way the algorithm goes. For example, here `log.trace("No replicants for %s", segment.getIdentifier())`